### PR TITLE
Secure backend containers

### LINKED ISSUE #57: Secure backend ports
By default PostgreSQL,  Redis, Mongo ports are listening on `0.0.0.0`. 
It should be restricted to private ip. 

For references:
https://github.com/kobotoolbox/kobo-docker/issues/249

### LINKED ISSUE #77: Firewall can block frontend containers from connecting to backend containers
@duvld said: 

make sure if running on Linux to either disable firewall or make a rule to allow kpi, if firewall blocks kpi from connecting to backend strange errors will come up.

Maybe we should write something about this in a readme or in the `run.py -s` prompts.

_Originally posted by @duvld in https://github.com/kobotoolbox/kobo-install/pull/76#issuecomment-602635151_

### LINKED ISSUE #79: [two-databases] KPI data is not copied from KC to KPI when using default options.
KPI database is created but empty.

### PR DESCRIPTION
**Pull request by noliveleger**
Closes #57
Closes #77
Closes #79 

- MongoDB and Redis with passwords and 
- Do not expose backend ports by default when not using multi servers environment.

This PR is related to 
- kobo-docker: https://github.com/kobotoolbox/kobo-docker/pull/280

### CODE REVIEW

File: à/.uniqid [hunk 1]
```diff
@@ -0,0 +1 @@
+1582213924
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply à/.uniqid]: Nope.

File: helpers/config.py [hunk 2]
```diff
@@ -876,13 +888,62 @@ def _round_nearest_quarter(dt):
...
+        mongo_root_username = CLI.get_response(
+            r"~^\w+$",
+            self.__config.get("mongo_root_username"))
+
+        CLI.colored_print("MongoDB root's password?", CLI.COLOR_SUCCESS)
+        mongo_root_password = CLI.get_response(
+            r"~^[{}]+$".format(self.REGEX_PASSWORD_ALLOWED_CHARACTERS),
+            self.__config.get("mongo_root_password"))
+
+        CLI.colored_print("MongoDB user's username?",
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply helpers/config.py]: @jnm It's already under `advanced options`.  No questions about MongoDB users are asked if `advanced options` is not chosen by the user at the beginning.

Do you want to an extra level of `advanced options` for Mongo? 
Something like:
```
- MongoDB user's username?
- MongoDB user's password?
- Do you want to change MongoDB root's credentials?
```

File: helpers/config.py [hunk 3]
```diff
@@ -1243,12 +1344,12 @@ def __questions_super_user_credentials(self):
...
                 not self.first_time:
-            CLI.colored_print("╔══════════════════════════════════════════════════════╗", CLI.COLOR_WARNING)
-            CLI.colored_print("║ Super user's password has been changed!              ║", CLI.COLOR_WARNING)
-            CLI.colored_print("║ Don't forget to apply these changes in Postgres too. ║", CLI.COLOR_WARNING)
-            CLI.colored_print("║ Super user's password won't be updated if the        ║", CLI.COLOR_WARNING)
-            CLI.colored_print("║ database already exists.                             ║", CLI.COLOR_WARNING)
-            CLI.colored_print("╚══════════════════════════════════════════════════════╝", CLI.COLOR_WARNING)
+            CLI.colored_print("╔═════════════════════════════════════════════════════════╗", CLI.COLOR_WARNING)
+            CLI.colored_print("║ Super user's password has been changed!                 ║", CLI.COLOR_WARNING)
+            CLI.colored_print("║ Do NOT forget to apply these changes in PostgreSQL too. ║", CLI.COLOR_WARNING)
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply helpers/config.py]: Nothing changed here. `psql` is not used.
It just writes the password in the config files like. You are right, it's used by KPI to create the super user.

https://github.com/kobotoolbox/kpi/blob/two-databases/docker/init.bash#L24-L25
```
echo 'Creating superuser...'
python manage.py create_kobo_superuser
```

But the script doesn't handle password update. That's why this message is displayed to user.

File: helpers/upgrading.py [hunk 4]
```diff
@@ -30,9 +31,8 @@ def _kpi_db_alias_kludge(command):
 
     kpi_run_command = ["docker-compose",
                        "-f", "docker-compose.frontend.yml",
-                       "-f", "docker-compose.frontend.override.yml"]
-    if config.get("docker_prefix", "") != "":
-        kpi_run_command += ["-p", config.get("docker_prefix")]
+                       "-f", "docker-compose.frontend.override.yml",
+                       "-p", config_object.get_prefix("frontend")]
     kpi_run_command += ["run", "--rm", "kpi"]
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply helpers/upgrading.py]: Indeed ;-)

File: helpers/config.py [hunk 5]
```diff
@@ -224,8 +221,25 @@ def generate_password(cls):
...
         :return: str
         """
-        return "".join(choice(cls.PASSWORD_ALLOWED_CHARACTERS)
-                       for x in range(randint(10, 16)))
+        characters = string.ascii_letters \
+                     + "!$%+-_^~@#{}[]()/\'\"`~,;:.<>" \
+                     + string.digits
+        required_chars_count = 12
+        stop = False
+        tmp_pwd = set()
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply helpers/config.py]: Good point. I'll remove the "uniqueness" of each character.

File: helpers/config.py [hunk 6]
```diff
@@ -850,13 +900,71 @@ def _round_nearest_quarter(dt):
...
+                              CLI.COLOR_WARNING)
+            CLI.colored_print("║ MongoDB root's and/or user's credentials have changed! ║",
+                              CLI.COLOR_WARNING)
+            CLI.colored_print("║ Do NOT forget to apply these changes in MongoDB too.   ║",
+                              CLI.COLOR_WARNING)
+            CLI.colored_print("║ Credentials will NOT be updated if the database has    ║",
+                              CLI.COLOR_WARNING)
+            CLI.colored_print("║ been previously created.                               ║",
+                              CLI.COLOR_WARNING)
+            CLI.colored_print("║ Visit https://tinyurl.com/s5488xp for more info.       ║",
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply helpers/config.py]: Ok, I'll take a look at it.

File: helpers/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jnm) [reply helpers/config.py]: Oh, okay. This seems fine the way it is.

File: helpers/config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jnm) [reply helpers/config.py]: Sorry, my complaint was with the message: "Do NOT forget to apply these changes **in PostgreSQL** too" (emphasis mine). I know that was there before, but it's probably confusing. Let's change it; I suggest:
> You have configured a new password for the super user. This change will \*not\* take effect if KoBoToolbox has ever been started before. Please use the web interface to change passwords for existing users. If you've forgotten your password:
> 1. Enter the KPI container with `./run.py -cf exec kpi bash`;
> 2. Create a new super user with `./manage.py createsuperuser`;
> 3. Type `exit` to leave the KPI container.

[the remark this answers is not in the document]
Comment (noliveleger) [reply helpers/config.py]: Way better ;-)

**Comment by jnm** [comment]
@noliveleger, thanks for the changes. Based on https://www.flowdock.com/app/kobotoolbox/kobo/threads/B31D-CYOT1-c8ocV07uEKueiDuP, the next order of business is:
- [ ] Respond with information about minimum (and maximum?) length if a value is invalid instead of "Sorry, I didn't understand that"
- [ ] Automate the setting of Mongo and Redis passwords so that user doesn't have to worry about it (or do it wrong)

Regarding the first point, I think it's fine to force people to change their too-short passwords so long as `kobo-install` can make the change take effect, i.e. provided that the script has a way to set database passwords after the first run of the application has already completed.

**Comment by almereyda** [comment]
Congratulations to the work done here. :clap:

**Pull request merged**